eth/vm/code_stream.py

```diff
@@ -14,27 +13,27 @@
 PUSH1, PUSH32, STOP = opcode_values.PUSH1, opcode_values.PUSH32, opcode_values.STOP
 
 
-class CodeStream(object):
-    stream = None
-    _length_cache = None
-    _raw_code_bytes = None
-    invalid_positions = None  # type: Set[int]
-    valid_positions = None  # type: Set[int]
+class CodeStream:
+    __slots__ = ['_length_cache', '_raw_code_bytes', 'invalid_positions', 'valid_positions', 'pc']
 
     logger = logging.getLogger('eth.vm.CodeStream')
 
     def __init__(self, code_bytes: bytes) -> None:
         validate_is_bytes(code_bytes, title="CodeStream bytes")
-        stream = io.BytesIO(code_bytes)
-        self.stream = stream
-        self._bound_stream_read = stream.read
+        self.pc = 0
         self._raw_code_bytes = code_bytes
         self._length_cache = len(code_bytes)
-        self.invalid_positions = set()
-        self.valid_positions = set()
+        self.invalid_positions = set()  # type: Set[int]
+        self.valid_positions = set()  # type: Set[int]
 
     def read(self, size: int) -> bytes:
-        return self._bound_stream_read(size)
+        old_pc = self.pc
+        target_pc = old_pc + size
+        if target_pc >= self._length_cache:
+            self.pc = self._length_cache
+        else:
+            self.pc = target_pc
+        return self._raw_code_bytes[old_pc:target_pc]
```

Comment:
Did you try using a `memoryview` here to avoid the copy?

Reply:
Nice, will try

Reply:
Wow, quite surprised that it doesn't speed things up. I stumbled on a 0.85% speedup in `push_XX` though (since switching to a `memoryview` broke things in there). Pushing...